dora/src/cpu/x64/asm.rs

```diff
@@ -393,13 +393,42 @@ pub fn emit_op(buf: &mut MacroAssembler, opcode: u8) {
     buf.emit_u8(opcode);
 }
 
+pub fn emit_vex3_rxbm(buf: &mut MacroAssembler, r: u8, x: u8, b: u8, m: u8) {
+    assert!(r == 0 || r == 1);
+    assert!(x == 0 || x == 1);
+    assert!(b == 0 || b == 1);
+    assert!(m > 1 && m < 4);
+
+    buf.emit_u8(0b11000100);
+    buf.emit_u8(!r << 7 | (!x << 6) & 0b1000000 | (!b << 5) & 0b100000 | m);
+}
+
+pub fn emit_vex3_wvlp(buf: &mut MacroAssembler, w: u8, v: u8, l: u8, p: u8) {
+    assert!(w == 0 || w == 1);
+    assert!(v < 16);
+    assert!(l == 0 || l == 1);
+    assert!(p < 4);
+
+    buf.emit_u8(w << 7 | (!v << 3) & 0b1111000 | l << 2 | p);
+}
+
+pub fn emit_vex2(buf: &mut MacroAssembler, r: u8, v: u8, l: u8, p: u8) {
+    assert!(r == 0 || r == 1);
+    assert!(v < 16);
+    assert!(l == 0 || l == 1);
+    assert!(p < 4);
+
+    buf.emit_u8(0b11000101);
+    buf.emit_u8(!r << 7 | (!v << 3) & 0b1111000 | l << 2 | p);
+}
+
 pub fn emit_rex(buf: &mut MacroAssembler, w: u8, r: u8, x: u8, b: u8) {
     assert!(w == 0 || w == 1);
     assert!(r == 0 || r == 1);
     assert!(x == 0 || x == 1);
     assert!(b == 0 || b == 1);
 
-    buf.emit_u8(0x4 << 4 | w << 3 | r << 2 | x << 1 | b);
+    buf.emit_u8(1 << 6 | w << 3 | r << 2 | x << 1 | b);
```

Comment:
I would propose to use `0x40` instead here, makes it easier to recognize this as rex-byte.